tensorflow/contrib/rnn/kernels/lstm_ops.h

```diff
@@ -0,0 +1,272 @@
+#ifndef TENSORFLOW_KERNELS_LSTM_OPS_H_
+#define TENSORFLOW_KERNELS_LSTM_OPS_H_
+
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+#include "tensorflow/core/framework/tensor_types.h"
+#include "tensorflow/core/platform/types.h"
+#include "tensorflow/stream_executor/stream.h"
+
+namespace tensorflow {
+
+#if GOOGLE_CUDA
+
+namespace {
+template <typename T>
+perftools::gputools::DeviceMemory<T> AsDeviceMemory(const T* cuda_memory) {
+  perftools::gputools::DeviceMemoryBase wrapped(const_cast<T*>(cuda_memory));
+  perftools::gputools::DeviceMemory<T> typed(wrapped);
+  return typed;
+}
+}  // namespace
+
+#endif  // GOOGLE_CUDA
+
+namespace functor {
+
+template <typename Device, bool USE_CUBLAS>
+struct LSTMCellBlockFprop {
+  void operator()(
+      perftools::gputools::Stream* stream, const Device& d,
+      const int batch_size, const int input_size, const int cell_size, const float forget_bias,
+      typename TTypes<float>::ConstMatrix x,
+      typename TTypes<float>::Matrix xh,
+      typename TTypes<float>::ConstMatrix states_prev,
+      typename TTypes<float>::ConstMatrix w,
+      typename TTypes<float>::ConstVec b,
+      typename TTypes<float>::Matrix h,
+      typename TTypes<float>::Matrix states) {
+    // Pointer offsets.
+    Eigen::array<int, 2> i_offsets  = {0, cell_size * 0};
+    Eigen::array<int, 2> cs_offsets = {0, cell_size * 1};
+    Eigen::array<int, 2> f_offsets  = {0, cell_size * 2};
+    Eigen::array<int, 2> o_offsets  = {0, cell_size * 3};
+    Eigen::array<int, 2> ci_offsets = {0, cell_size * 4};
+    Eigen::array<int, 2> co_offsets = {0, cell_size * 5};
+    Eigen::array<int, 2> h_offsets  = {0, cell_size * 6};
+
+    Eigen::array<int, 2> cell_extents = {batch_size, cell_size};
+
+    // xh = [x, h_prev].
+    Eigen::array<int, 2> xh_x_offsets = {0, 0};
+    Eigen::array<int, 2> xh_x_extents = {batch_size, input_size};
+    Eigen::array<int, 2> xh_h_offsets = {0, input_size};
+    Eigen::array<int, 2> xh_h_extents = {batch_size, cell_size};
+
+    auto h_prev =
+        states_prev.slice(h_offsets, cell_extents);
+
+    xh.slice(xh_x_offsets, xh_x_extents).device(d) = x;
+    xh.slice(xh_h_offsets, xh_h_extents).device(d) = h_prev;
+
+    Eigen::array<int, 2> states_offsets = {0, 0};
+    Eigen::array<int, 2> states_extents = {batch_size, cell_size * 4};
+
+    // states = xh * w + b
+    if (USE_CUBLAS) {
```

Comment:
Btw I think before we'd move this into core, this would need cleanup -- there's way too many #if GOOGLE_CUDAs here I think.